mountpoint-s3/src/checksums/block.rs

```diff
@@ -0,0 +1,211 @@
+    fn from(value: ChecksummedBlock) -> Self {
+        value.into_checksummed_bytes()
+    }
+}
+
+// Implement equality for tests only. We implement equality, and will panic if the data is corrupted.
+#[cfg(test)]
+impl PartialEq for ChecksummedBlock {
+    fn eq(&self, other: &Self) -> bool {
+        if self.bytes != other.bytes {
+            return false;
+        }
+
+        if self.checksum == other.checksum {
+            return true;
+        }
+
+        self.validate().expect("should be valid");
+        other.validate().expect("should be valid");
+
+        true
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use bytes::Bytes;
+    use mountpoint_s3_crt::checksums::crc32c;
+
+    use super::*;
+
+    #[test]
+    fn test_into_bytes() {
+        let bytes = Bytes::from_static(b"some bytes");
+        let expected = bytes.clone();
+        let checksum = crc32c::checksum(&bytes);
+        let checksummed_block = ChecksummedBlock::new(bytes, checksum);
+
+        let actual = checksummed_block.into_bytes().unwrap();
+        assert_eq!(expected, actual);
+    }
+
+    #[test]
+    fn test_into_bytes_integrity_error() {
+        let bytes = Bytes::from_static(b"some bytes");
+        let checksum = crc32c::checksum(&bytes);
+        let mut checksummed_block = ChecksummedBlock::new(bytes, checksum);
+        checksummed_block.bytes = Bytes::from_static(b"new bytes");
+
+        let actual = checksummed_block.into_bytes();
+        assert!(matches!(actual, Err(IntegrityError::ChecksumMismatch(_, _))));
+    }
+
+    #[test]
+    fn test_into_checksummed_bytes() {
+        let bytes = Bytes::from_static(b"some bytes");
+        let checksum = crc32c::checksum(&bytes);
+        let checksummed_block = ChecksummedBlock::new(bytes, checksum);
+        let checksummed_bytes = checksummed_block.clone().into_checksummed_bytes();
+
+        assert_eq!(
+            checksummed_block.into_bytes().unwrap(),
+            checksummed_bytes.into_bytes().unwrap()
+        );
+    }
+
+    #[test]
+    fn test_extend() {
+        let bytes = Bytes::from_static(b"some bytes");
+        let expected = Bytes::from_static(b"some bytes extended");
+        let checksum = crc32c::checksum(&bytes);
+        let mut checksummed_block = ChecksummedBlock::new(bytes, checksum);
+
+        let extend = Bytes::from_static(b" extended");
+        let extend_checksum = crc32c::checksum(&extend);
+        let extend = ChecksummedBlock::new(extend, extend_checksum);
+        checksummed_block.extend(extend);
+        let actual = checksummed_block.bytes;
+        assert_eq!(expected, actual);
+    }
+
+    #[test]
+    fn test_extend_self_corrupted() {
+        let bytes = Bytes::from_static(b"some bytes");
+        let checksum = crc32c::checksum(&bytes);
+        let mut checksummed_block = ChecksummedBlock::new(bytes, checksum);
+
+        let currupted_bytes = Bytes::from_static(b"corrupted data");
+        checksummed_block.bytes = currupted_bytes.clone();
```

Comment:
nit: We can just put the corrupted bytes straight into the `new` call? Might be a bit clearer

alternatively, we also have `from_bytes` so we can make these tests a bit simpler